usr/lib/xtable/luaxt.c

```diff
@@ -0,0 +1,203 @@
+	lua_pop(L, 1); /* table */
+
+	if (lua_pcall(L, nargs, nresults, 0) != LUA_OK) {
+		pr_err("Failed to call Lua function %s: %s\n", func_name,
+				lua_tostring(L, -1));
+		goto restore;
+	}
+
+	if (nresults == 1)
+		ret = lua_toboolean(L, -1);
+
+restore:
+	lua_settop(L, base);
+	return ret;
+}
+
+#define LIBXTABLE_HELPER_CB(hook)			   	\
+static void luaxt_##hook##_help(void)		   	\
+{							   	\
+	luaxt_run(L, "help", "luaxt_"#hook, 0, 0);	\
+}
+
+#define LIBXTABLE_INITER_CB(hook)				   	\
+static void luaxt_##hook##_init(struct xt_entry_##hook *hook)   	\
+{								   	\
+	luaxt_run(L, "init", "luaxt_"#hook, 1, 0);		\
+}
+
+#define LIBXTABLE_PARSER_CB(hook)			   		\
+static int luaxt_##hook##_parse(int c, char **argv, int invert, unsigned int *flags,	\
+					const void *entry, struct xt_entry_##hook **hook)	\
+{							   		\
+	return luaxt_run(L, "parse", "luaxt_"#hook, 0, 1);	\
+}
+
+#define LIBXTABLE__FINALCHECKER_CB(hook)				\
+static void luaxt_##hook##_finalcheck(unsigned int flags)		\
+{							  		\
+	luaxt_run(L, "final_check", "luaxt_"#hook, 0, 0);	\
+}
+
+#define LIBXTABLE_PRINTER_CB(hook)  					\
+static void luaxt_##hook##_print(const void *entry, const struct xt_entry_##hook *hook, int numeric)	\
+{							   		\
+	luaxt_run(L, "print", "luaxt_"#hook, 0, 0);	\
+}
+
+#define LIBXTABLE_SAVER_CB(hook)				\
+static void luaxt_##hook##_save(const void *entry, const struct xt_entry_##hook *hook)  \
+{								\
+	luaxt_run(L, "save", "luaxt_"#hook, 0, 0);	\
+}
+
+#define LIBXTABLE_NEWCB(hook)  		\
+	LIBXTABLE_HELPER_CB(hook)   	\
+	LIBXTABLE_INITER_CB(hook)   	\
+	LIBXTABLE_PARSER_CB(hook)   	\
+	LIBXTABLE__FINALCHECKER_CB(hook)	\
+	LIBXTABLE_PRINTER_CB(hook)  	\
+	LIBXTABLE_SAVER_CB(hook)
+
+LIBXTABLE_NEWCB(match);
+LIBXTABLE_NEWCB(target);
+
+
+#define LIBXTABLE_NEWREG(hook)					\
+static struct xtables_##hook luaxt_##hook##_reg = {		\
+	.version = XTABLES_VERSION,				\
+	.name = LUAXTABLE_MODULE,					\
+	.size = XT_ALIGN(sizeof(luaxtable_info_t)),		\
+	.userspacesize = 0,					\
+	.help = luaxt_##hook##_help,			\
+	.init = luaxt_##hook##_init,			\
+	.parse = luaxt_##hook##_parse,		 	\
+	.final_check = luaxt_##hook##_finalcheck,	  	\
+	.print = luaxt_##hook##_print,		  	\
+	.save = luaxt_##hook##_save			 	\
+}
+
+LIBXTABLE_NEWREG(match);
+LIBXTABLE_NEWREG(target);
+
+static inline int luaxt_checkint(lua_State *L, int idx, const char *key)
+{
+	if (lua_getfield(L, idx, key) != LUA_TNUMBER) {
+		pr_err("invalid \'%s\' in match\n", key);
+		return 0;
+	}
+	int ret = lua_tointeger(L, -1);
+	lua_pop(L, 1);
+	return ret;
+}
+
+#define LIBXTABLE_LIB_CB(hook)				\
+static int luaxt_##hook(lua_State *L)		\
+{							\
+	luaL_checktype(L, 1, LUA_TTABLE);		\
+							\
+	luaxt_##hook##_reg.revision = luaxt_checkint(L, 1, "revision");		\
+	luaxt_##hook##_reg.family = luaxt_checkint(L, 1, "family");		\
+	xtables_register_##hook(&luaxt_##hook##_reg);		\
+	lua_pushvalue(L, 1);						\
+	lua_rawsetp(L, LUA_REGISTRYINDEX, "luaxt_"#hook);		\
```

Comment:
you don't need to use a string here.. you can just use `luaxt_##hook` and use a `void *` as key.. ;-)